inmoose/limma/ebayes.py

```diff
@@ -19,9 +19,282 @@
+          :code:`df_prior + df_residual` or :code:`sum(df_residual)`, whichever
+          is smaller.
+        - :code:`s2_post`: row-wise array giving the posterior values for
+          :code:`sigma**2`
+        - :code:`var_prior`: column-wise array giving estimated prior values for
+          the variance of the log2-fold-changes for differentially expressed
+          gene for each contrast. Used for evaluating :code:`lods`.
+        - :code:`F`: row-wise array of moderated *F*-statistics for testing all
+          contrasts defined by the columns of :code:`fit` simultaneously equal
+          to zero.
+        - :code:`F_p_value`: row-wise array giving *p*-values corresponding to
+          :code:`F`
+
+        The matrices :code:`t`, :code:`p_value`, :code:`lods` have the same
+        dimensions as the input object :code:`fit`, with rows corresponding to
+        genes and columns to coefficients or constrats.  The vector
+        :code:`s2_prior`, :code:`df_prior`, :code:`df_total`, :code:`F` and
+        :code:`F_p_value` correspond to rows, with length equal to the number
+        of genes. The vector :code:`var_prior` corresponds to columns, with
+        length equal to the number of contrasts. If :code:`s2_prior` or
+        :code:`df_prior` have length 1, then the same value applies to all
+        genes.
+
+        :code:`s2_prior`, :code:`df_prior` and :code:`var_prior` contain
+        empirical Bayes hyperparameters used to obtain :code:`df_total`,
+        :code:`s2_post` and :code:`lods`.
+    """
+    if not isinstance(fit, MArrayLM):
+        raise ValueError("fit is not a valid MArrayLM object")
+    if isinstance(trend, bool) and trend and (fit.Amean is None):
+        raise ValueError("Need Amean component in fit to estimate trend")
+    eb = _ebayes(
+        fit=fit,
+        proportion=proportion,
+        stdev_coef_lim=stdev_coef_lim,
+        trend=trend,
+        robust=robust,
+        winsor_tail_p=winsor_tail_p,
+    )
+    fit.df_prior = eb["df_prior"]
+    fit.s2_prior = eb["s2_prior"]
+    fit.var_prior = eb["var_prior"]
+    fit.proportion = proportion
+    fit.s2_post = eb["s2_post"]
+    fit.t = eb["t"]
+    fit.df_total = eb["df_total"]
+    fit.p_value = eb["p_value"]
+    fit.lods = eb["lods"]
+    if (fit.design is not None) and np.linalg.matrix_rank(
+        fit.design
+    ) == fit.design.shape[1]:
+        F_stat = classifyTestsF(fit, fstat_only=True)
+        fit.F = np.asarray(F_stat)
+        df1 = F_stat.df1
+        df2 = F_stat.df2
+        # if Y ~ F(df1, df2), then df1 Y ~ chi2(df1) when df2 -> infinity
+        if df2 > 1e10:
+            fit.F_p_value = scipy.stats.chi2.sf(df1 * fit.F, df=df1)
+        else:
+            fit.F_p_value = scipy.stats.f.sf(fit.F, dfn=df1, dfd=df2)
+
+    return fit
+
+
+def _ebayes(
+    fit,
+    proportion=0.01,
+    stdev_coef_lim=(0.1, 4),
+    trend=False,
+    robust=False,
+    winsor_tail_p=(0.05, 0.1),
```

Comment:
the type of winsor_tail_p changes between the two functions

Reply:
hawk eye
fixed